config/routes.rb

```diff
@@ -45,6 +45,10 @@
   # Routes for GOV.UK Account discovery
   if ENV["FEATURE_FLAG_ACCOUNTS"] == "enabled"
     get "/transition-check/save-your-results" => "brexit_checker#save_results", as: :transition_checker_save_results
+
+    constraints subdomain: Services.accounts_api.gsub(".gov.uk", "") do
+      post "/", to: proc { [200, {}, [""]] }, as: :govuk_account_signup
```

Comment:
I'm having a bit of trouble understanding what this route and the constraint is for.   Can you explain what it will be used for?

Also, when I run rspec with `FEATURE_FLAG_ACCOUNTS=true`, I get the following error, which goes away when I remove lines 49-51 on `routes.rb`:
```
Failure/Error: require File.expand_path("../config/environment", __dir__)

FrozenError:
  can't modify frozen Array
# ./config/environment.rb:5:in `<top (required)>'
# ./spec/spec_helper.rb:9:in `<top (required)>'
# ./spec/validators/param_validator_spec.rb:1:in `<top (required)>'
```

Reply:
Sure, this line is a problem and needs working on, but perhaps easiest to understand if I separate into two different things I'm trying to do wtih it:

**Route to an external app, with a subdomain.**

I want a route I can call throughout the app `govuk_account_signup` that i could use with `_path` or `_url` so if we move the account server around  a bit we don't have to go back and change a bunch of views.

For this I nicked inspiration from what we do for [email_frontend_signup](https://github.com/alphagov/finder-frontend/blob/master/config/routes.rb#L27). It seems that if you can't route to a controller you can pass this Proc object (which is a new one to me!). That allows us to post.

However in testing i seemed to be running into an assumption that our app would be on the same domain. That's what's going on with the email signup, it's posting across to a fellow app.

Ours is going to be on paas right? With a different subdomain?
I looked up how to do that and [came across constraints](https://guides.rubyonrails.org/routing.html#request-based-constraints). These seem to be the best way to route into a subdmain like `login.service...`.

Perhaps that's not the pattern I should have reached for here though?

**Have the correct domain taken from Plek**

Trouble is I want the app to route correctly for develoment, integration, staging and tests. So i figure what this domain is (login.service.staging.gov.uk?), needs to come out of a plek evironment variable. For that i've got `Services.accounts_api`. However plek would come with the gov.uk domain on the end which i don't want so i gsub.

This doesn't seem like a good pattern to me, and is causing issues. Going to rearch a plan B now, but any better ideas would be appreciated.

**can't modify frozen Array**

Yeah this is the core problem i'm having in tests. If I drop a binding.pry into routes it seems that the rotues file only gets loaded once during the tests...

I'd wanted it so the routes wouldn't work unless the feature flag was set... but if these aren't getting rechecked that probably won't work.

If i drop the conditional however, then the routes makes a call to `Services.account_api`. Problem being I've also put this behind [a feature flag gate](https://github.com/alphagov/finder-frontend/pull/2207/files#diff-ee054dafd7eadee4c8baf208077209baR34-R38), because i wanted to use `Env.fetch` to give a better error message if the feature flag was set to enabled but the required additional Env vars weren't there. This seemed like a reasonable saftey feature but it's now causing problems in these tests :thinking:

Reply:
Thanks for the explanation, I understand now. :smile:

I think there should be a simple solution to this.

If we define `PLEK_SERVICE_ACCOUNT_MANAGER_URI` in puppet (and `govuk-docker` for local development) to equal the account manager URL, you can get the URL in `finder-frontend` using `Plek.find("account-manager")`.  You can set a different value in each environment (dev, integration, staging and production) and the `.gov.uk` won't be appended (that only gets added if there is no override environment variable).

In `app/lib/services.rb`, you could have:
```
  if Rails.configuration.feature_flag_govuk_accounts
    def self.accounts_api
      Plek.find("account-manager")
    end
  end
```

And then in the view, just use `<%= form_tag Services.accounts_api .......`.

That should allow you to remove these lines from the routes and means we can change the URL of the account manager by updating the Plek environment variable in `govuk-docker` and puppet.